[ZIP-13] Transport Re-architecture for Pull Messages Between Seed Nodes (#52)
Zilliqa/ZIP: zips/zip-13.md

### Implementation Considerations

In terms of changes, following nodes will be affected.

#### Seed Node(client)
1) Seed node will always connect to `33135` for pull messages. For push messages like sending transactions and blacklisting will be handled at `33133` port.
2) For sending a message to source seed for pull messages , seed node will use startbyte `START_BYTE_SEED_TO_SEED_REQUEST` during `SendMessage()` function call.
3) In addition there will be a time-out for response message which will be initialized to `SEED_SYNC_LARGE_PULL_INTERVAL` during bufferevent creation through api `bufferevent_set_timeouts`.This is required as it is expected that for some messages, there will be no response.In order to cleanup the bufferevent at sender side, we have to set timeout for messages or wait for timer expiration at application level and then do the cleanup.
4) In order to keep the event loop running during creation, it is required to add a timer event and the call `event_base_dispatch()`.If it's not done, the event loop terminates immediately and kills the main thread after creation.
5)Main thread is running the event loop and responsible for handling network events.

> 

Reviewer: ```suggestion
```

#### Source Seed Node(server)

> 1) Server node will open another port on `33135` and the events for it will be dispatched to the same event loop handling listening events on `33133`.

Reviewer: ```suggestion
1) Server node will open another port on `33135` and the events for it will be dispatched to the same event loop handling listening events on port `33133`.
```

2) All messages with startByteType  `START_BYTE_SEED_TO_SEED_REQUEST` will be handled on new accept/read/event callbacks and rest all are dropped.

> 2) Server node will store the bufferevent for the request in `bufferEventMap`.The key of the map would be key = `IP:PORT`, value = `bufferevent*`

Reviewer: ```suggestion
3) Server node will store the bufferevent for the request in `bufferEventMap`. The key of the map would be key = `IP:PORT`, value = `bufferevent*`
```

> 3) For sending the response server node will use `START_BYTE_SEED_TO_SEED_RESPONSE` start byte in `SendMessage()` function call.

Reviewer: ```suggestion
4) For sending the response server node will use `START_BYTE_SEED_TO_SEED_RESPONSE` start byte in `SendMessage()` function call.
```

> 4) In case server does not send any response for incoming request, server node will just remove the bufferevent from bufferEventMap.Termination of socket connection will only be done by events from client side.

Reviewer: ```suggestion
4) In case server does not send any response for incoming request, the server node will just remove the bufferevent from bufferEventMap. Termination of socket connection will only be done by events from client side.
```

> 5) Buffer event cleanup will only be done by events from client side(socket closure,socket error).

Reviewer: ```suggestion
6) Buffer event cleanup will only be done by events from client side (socket closure and socket error).
```
